File: sdks/go/engine_version.go
package hatchet

import (
	"context"
	"fmt"
	"strconv"
	"strings"
)

// MinEngineVersion defines minimum engine versions for feature support.
var MinEngineVersion = struct {
	SlotConfig      string
	DurableEviction string
	Observability   string
}{
	SlotConfig:      "v0.78.23",
	DurableEviction: "v0.80.0",
	Observability:   "v0.82.0",
}

// GetEngineVersion retrieves the engine version from the server.
func (c *Client) GetEngineVersion(ctx context.Context) (string, error) {
	version, err := c.legacyClient.Dispatcher().GetVersion(ctx)
	if err != nil {
		return "", fmt.Errorf("failed to get engine version: %w", err)
	}
	return version, nil
}

// SupportsDurableEviction checks whether the engine version supports durable eviction.
func SupportsDurableEviction(engineVersion string) (bool, error) {
	return !semverLessThan(engineVersion, MinEngineVersion.DurableEviction), nil
}

// parseSemver parses a semver string like "v0.80.0" into (major, minor, patch).
func parseSemver(v string) (int, int, int, error) {
	v = strings.TrimPrefix(v, "v")
	parts := strings.SplitN(v, ".", 3)
	if len(parts) != 3 {
		return 0, 0, 0, fmt.Errorf("invalid semver: %s", v)
	}

	// Handle pre-release suffix in patch (e.g. "0-alpha.1")
	patchStr := strings.SplitN(parts[2], "-", 2)[0]

	major, err := strconv.Atoi(parts[0])
	if err != nil {
		return 0, 0, 0, fmt.Errorf("invalid semver major: %s", parts[0])
	}
	minor, err := strconv.Atoi(parts[1])
	if err != nil {
		return 0, 0, 0, fmt.Errorf("invalid semver minor: %s", parts[1])
	}
	patch, err := strconv.Atoi(patchStr)
	if err != nil {
		return 0, 0, 0, fmt.Errorf("invalid semver patch: %s", patchStr)
	}
	return major, minor, patch, nil
}

// semverLessThan returns true if version a is strictly less than version b.
func semverLessThan(a, b string) bool {
	aMaj, aMin, aPatch, err := parseSemver(a)
	if err != nil {
		return false
	}
	bMaj, bMin, bPatch, err := parseSemver(b)
	if err != nil {
		return false
	}
	if aMaj != bMaj {
		return aMaj < bMaj
	}
	if aMin != bMin {
		return aMin < bMin
	}
	return aPatch < bPatch
}

File: sdks/ruby/src/lib/hatchet/engine_version.rb
# frozen_string_literal: true

module Hatchet
  # Minimum engine versions required for specific SDK features.
  #
  # Mirrors :class:`hatchet_sdk.engine_version.MinEngineVersion` in the
  # Python SDK.
  module MinEngineVersion
    SLOT_CONFIG = "v0.78.23"
    DURABLE_EVICTION = "v0.80.0"
    OBSERVABILITY = "v0.82.0"
  end

  # Semver parsing + comparison helpers used to gate features by engine version.
  module EngineVersion
    module_function

    # Parse a semver string like ``"v0.78.23"`` into ``[major, minor, patch]``.
    #
    # Returns ``[0, 0, 0]`` if parsing fails, matching the Python helper.
    #
    # @param version [String, nil] The version string (with or without a ``v`` prefix, optional ``-pre`` suffix)
    # @return [Array(Integer, Integer, Integer)]
    def parse_semver(version)
      return [0, 0, 0] if version.nil?

      v = version.to_s
      v = v.sub(/\Av/, "")
      v = v.split("-", 2).first || ""

      parts = v.split(".")
      return [0, 0, 0] if parts.length != 3

      [Integer(parts[0]), Integer(parts[1]), Integer(parts[2])]
    rescue ArgumentError, TypeError
      [0, 0, 0]
    end

    # @param left_version [String, nil]
    # @param right_version [String, nil]
    # @return [Boolean] true if semver ``left_version`` is strictly less than ``right_version``
    def semver_less_than?(left_version, right_version)
      (parse_semver(left_version) <=> parse_semver(right_version)).negative?
    end

    class << self
      alias semver_less_than semver_less_than?
    end
  end
end
